File: lib/domain/crop/interactors/crop_detail_interactor.rb
# frozen_string_literal: true

module Domain
  module Crop
    module Interactors
      class CropDetailInteractor < Domain::Crop::Ports::CropDetailInputPort
        def initialize(output_port:, user_id:, show_detail_read_gateway:, user_lookup:)
          @output_port = output_port
          @show_detail_read_gateway = show_detail_read_gateway
          @user_id = user_id
          @user_lookup = user_lookup
        end

        def call(crop_id)
          user = @user_lookup.find(@user_id)
          access_filter = Domain::Shared::Policies::CropPolicy.record_access_filter(user)
          show_detail_snapshot = @show_detail_read_gateway.find_show_detail_snapshot(crop_id: crop_id)
          crop_detail_dto = Mappers::CropShowDetailMapper.from_snapshot(show_detail_snapshot)
          Domain::Shared::ReferenceRecordAuthorization.assert_view_allowed!(access_filter, crop_detail_dto.crop)
          @output_port.on_success(crop_detail_dto)
        rescue Domain::Shared::Policies::PolicyPermissionDenied => e
          @output_port.on_failure(e)
        rescue Domain::Shared::Exceptions::RecordNotFound => e
          @output_port.on_failure(Domain::Shared::Dtos::Error.new(e.message))
        rescue Domain::Shared::Exceptions::RecordInvalid => e
          @output_port.on_failure(Domain::Shared::Dtos::Error.new(e.message))
        end
      end
    end
  end
end

File: crates/agrr-domain/src/crop/interactors/crop_detail_interactor.rs
//! Ruby: `Domain::Crop::Interactors::CropDetailInteractor`

use crate::crop::dtos::CropDetailOutput;
use crate::crop::gateways::CropGateway;
use crate::crop::ports::{CropDetailOutputPort, DetailFailure};
use crate::shared::dtos::Error;
use crate::shared::exceptions::{RecordInvalidError, RecordNotFoundError};
use crate::shared::gateways::UserLookupGateway;
use crate::shared::policies::crop_policy;
use crate::shared::reference_record_authorization;

pub struct CropDetailInteractor<'a, G, O, U> {
    output_port: &'a mut O,
    gateway: &'a G,
    user_id: i64,
    user_lookup: &'a U,
}

impl<'a, G, O, U> CropDetailInteractor<'a, G, O, U>
where
    G: CropGateway,
    O: CropDetailOutputPort,
    U: UserLookupGateway,
{
    pub fn new(output_port: &'a mut O, user_id: i64, gateway: &'a G, user_lookup: &'a U) -> Self {
        Self {
            output_port,
            gateway,
            user_id,
            user_lookup,
        }
    }

    pub fn call(
        &mut self,
        crop_id: i64,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let user = self.user_lookup.find(self.user_id);
        let access_filter = crop_policy::record_access_filter(user);

        let crop_detail = match self.gateway.find_crop_show_detail(crop_id) {
            Ok(detail) => detail,
            Err(err) => {
                if err.downcast_ref::<RecordNotFoundError>().is_some()
                    || err.downcast_ref::<RecordInvalidError>().is_some()
                {
                    self.output_port
                        .on_failure(DetailFailure::Error(Error::new(err.to_string())));
                    return Ok(());
                }
                return Err(err);
            }
        };

        if let Err(policy) =
            reference_record_authorization::assert_view_allowed(&access_filter, &crop_detail.crop)
        {
            self.output_port.on_failure(DetailFailure::Policy(policy));
            return Ok(());
        }

        self.output_port
            .on_success(CropDetailOutput::new(crop_detail.crop));
        Ok(())
    }
}

#[cfg(test)]
mod interactors_crop_detail_interactor_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/crop/interactors_crop_detail_interactor_test.rs"));
}
